[export] Skip guard propagation for export only.

zhxchen17 opened the pull request:
Summary:
IIUC, from the beginning of torch.export to the current moment, we never really find a way to meaningfully use generated guards from torch dynamo, since export by design must output a non Python dependent artifact to users. Therefore I think in the forseeable future we still don't need dynamo guards to be present.

Recently we ran into some slowness in dynamo for internal models and observed a substantial time consumed in guard propagation logic, especially when we're dealing with list VTs. This PR effectively detects whether we're exporting and if so skip the entire guard propagation logic.

This doesn't resolve the fundemental issue of O(n^2) algorithm used in VT but greatly reduce the constant factor of it, thus still gives us great speedup for exporting time. We've discussed the resolution in pt2 core meeting but as a quick unblocker in the short term I think we could also try to skip guards to unblock several internal models.

Some experiment ran on the internal FM model:
local_fm (training): 15min -> 4min
ads launcher ir generation (training, D50847630): >1hr -> 30min

Test Plan: CI

Differential Revision: D50914819




cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

ezyang commented on torch/_dynamo/variables/base.py:
```diff
@@ -146,6 +148,13 @@ def propagate(*vars: List[List["VariableTracker"]]):
+            "guards": guards,
+        }
+        tcx = TracingContext.get()
+        if tcx is not None and tcx.export:
```
This needs a big commentary describing what the idea here is

zhxchen17 committed:
[export] Skip guard propagation for export only. (#112685)

Summary:

IIUC, from the beginning of torch.export to the current moment, we never really find a way to meaningfully use generated guards from torch dynamo, since export by design must output a non Python dependent artifact to users. Therefore I think in the forseeable future we still don't need dynamo guards to be present.

Recently we ran into some slowness in dynamo for internal models and observed a substantial time consumed in guard propagation logic, especially when we're dealing with list VTs. This PR effectively detects whether we're exporting and if so skip the entire guard propagation logic.

This doesn't resolve the fundemental issue of O(n^2) algorithm used in VT but greatly reduce the constant factor of it, thus still gives us great speedup for exporting time. We've discussed the resolution in pt2 core meeting but as a quick unblocker in the short term I think we could also try to skip guards to unblock several internal models.

Some experiment ran on the internal FM model:
local_fm (training): 15min -> 4min
ads launcher ir generation (training, D50847630): >1hr -> 30min

Test Plan: CI

Reviewed By: tugsbayasgalan

Differential Revision: D50914819

zhxchen17 commented:
Added some comments inline about why we skip guards propagation for export.

jansel reviewed (changes requested):
I think this will be made obsolete by:
https://github.com/pytorch/pytorch/pull/111726

Does that PR fix the performance issues you are seeing?  If so maybe we don't need to add an extra mode.

zhxchen17 commented:
> I think this will be made obsolete by: #111726
> 
> Does that PR fix the performance issues you are seeing? If so maybe we don't need to add an extra mode.

I will do a test with that PR.
